A16-tcp-user-timeout.md

```diff
@@ -0,0 +1,68 @@
+TCP User Timeout
+----
+* Author(s): Yash Tibrewal
+* Approver:
+* Status: Draft
+* Implemented in:
+* Last updated: 2018-8-22
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/6VZYCFZpyTI
+
+## Abstract
+
+Provide an option to set the socket option, TCP_USER_TIMEOUT on TCP sockets for
+Linux kernels 2.6.37 and later.
+
+## Background
+
+TCP_USER_TIMEOUT is a socket option that has been available on Linux since Linux
+2.6.37. As described the Linux manual page, "This option takes an unsigned int
+as an argument. When the value is greater than 0, it specifies the maximum
+amount of time in milliseconds that transmitted data may remain unacknowledged
+before TCP will forcibly close the corresponding connection and return ETIMEDOUT
+to the application.  If the option value is specified as 0, TCP will to use the
+system default."
+
+At the time this proposal was written, gRPC Core faced an issue where data could
+be stuck for a very long time at the gRPC TCP layer. For cases where the network
+goes down, and the socket cannot accept more data for writing, an RPC could be
+blocked for a very long time. Even if the deadline expired on such an RPC, gRPC
+Core was incapable of shutting down the RPC, because the TCP layer would stil
+have references to user data. Setting TCP_USER_TIMEOUT to a reasonable value
```

Comment:
Worth noting that fixing the reference issue would come at the cost of performance.